### PR TITLE
extmod/modbluetooth: Add pairing/bonding support.

### PR DESCRIPTION
**Pull request by jimmo**
WIP, looking for feedback on the approach.

Much of this work was done by @andrewleech in #6289, #6312, and #6447. This PR changes:
 - Moved the bonding secret store into Python (via _IRQ_GET_SECRET and _IRQ_SET_SECRET events).
 - Adds characteristic flags to enable auto-pairing (e.g. READ_ENCRYPTED)
 - Adds passkey authentication.
 - gap_pair just takes conn_handle, and pairing/bonding parameters are set via config().
 - Documentation

Currently only supported on NimBLE on STM32 and Unix, as it requires synchronous (non-ringbuffer) events. Enabled via `MICROPY_PY_BLUETOOTH_ENABLE_PAIRING_BONDING` macro (which is now enabled by default for NimBLE STM32/Unix). Some BlueKitchen support is implemented, but will need to wait until it's migrated to synchronous events. The secret store is designed to be compatible with BlueKitchen's TLV store API.

Tested against Android and BlueZ.

This PR adds:
 - Better error handling for oversize characteristic reads.
 - An _IRQ_CONNECTION_UPDATE event. Not specifically related to pairing and bonding, but useful information about the connection state.
 - An _IRQ_ENCRYPTION_UPDATE event to get notified about pairing/bonding state change.
 - Widens characteristic flags to 16-bit and allows for setting encryption/authentication/authorisation requirements.
 - Allows the user to fail the read request IRQ with a custom error code.
 - Configuration of pairing/bonding parameters (le_secure, bond, mitm, io)
 - gap_pair() method to initiate pairing.
 - _IRQ_GET_SECRET and IRQ_SET_SECRET events for persistent key management.
 - Simplify the handing of numeric arguments in invoke_irq_handler (used in synchronous events).
 - An _IRQ_PASSKEY_ACTION event to be notified of an authentication request.
 - gap_passkey() method to respond to an authentication request.
 - ble_bonding_peripheral.py example, showing implementation of key persistence and authentication. (WIP)

Four major outstanding pieces:
 - Write multi-test.
 - Support address resolving in central mode (perhaps via a _IRQ_RESOLVED_SCAN_RESULT event)
 - Improve handing of RPA mode. NimBLE doesn't seem to support using a non-fixed IRK (see https://github.com/apache/mynewt-nimble/issues/887)
 - PYBD doesn't work with RPA mode. (Works fine on Unix+Zephyr and WB55)

### CODE REVIEW

**Comment by dmazzella** [comment]
+1 for bonding secret store into Python

**Comment by andrewleech** [comment]
Minor note: in top post `MICROPY_BLUETOOTH_ENABLE_PAIRING_BONDING` should be `MICROPY_PY_BLUETOOTH_ENABLE_PAIRING_BONDING`

File: extmod/nimble/modbluetooth_nimble.c [hunk 1]
```diff
@@ -808,6 +889,40 @@ int mp_bluetooth_set_preferred_mtu(uint16_t mtu) {
...
+    switch (action) {
+        case MP_BLUETOOTH_PASSKEY_ACTION_INPUT: {
+            io.passkey = passkey;
+            break;
+        }
+        case MP_BLUETOOTH_PASSKEY_ACTION_DISP: {
+            io.passkey = passkey;
+            break;
+        }
+        case MP_BLUETOOTH_PASSKEY_ACTION_NUMCMP: {
```
Comment (andrewleech) [inline extmod/nimble/modbluetooth_nimble.c:910]: `MP_BLUETOOTH_PASSKEY_ACTION_DISP` above and `MP_BLUETOOTH_PASSKEY_ACTION_NUMCMP` don't match the defines

**Comment by jimmo** [comment]
> Minor note: in top post `MICROPY_BLUETOOTH_ENABLE_PAIRING_BONDING` should be `MICROPY_PY_BLUETOOTH_ENABLE_PAIRING_BONDING`

Fixed. (And the PASSKEY defines)

**Comment by andrewleech** [comment]
My testing of this PR today has gone very well, though I think the IRK issue can cause some strange behaviors if there are multiple devices running. 

I've been testing primarily android central connecting to micropython peripheral on both unix (zephyr hci radio) and stm32 (mynewt/nimble hci radio).

Pairing and bonding works from both explicit pairing request and/or reading a characteristic with the `_FLAG_READ_ENCRYPTED` attribute.

This is with both the `ble_bonding_peripheral.py` example which uses a pin to pair (recognized as authenticated) and my custom application which has no pin (recognized as not authenticated). The link is encrypted and bonded in both cases.

The bond is maintained reliably across power cycles of either end of the link.

If the ble is disabled on micropython: `ble.active(False)` then enabled again `ble.active(True)` the bond doesn't seem to be persisted correctly, however I expect this is a pretty uncommon use case at this stage.

**Comment by jimmo** [comment]
I've added a commit that adds a workaround for the fixed IRK.

**Comment by andrewleech** [comment]
I've continued testing, still with micropython running as peripheral but a different Android (samsung A20, Android 10) phone running central (both nrf-connect and custom app).

With this phone, the pairing is not persisting over the micropython reboot when running on my custom stm32 application.
I can connect/disconnect/connect the phone while the stm32 is running just fine (pairing works) but after rebooting stm32 and re-connecting the pairing does not re-establish.

This exact same procedure with my previous Android phone (Xiaomi MI6 running Lineage OS Android 10) works correctly - pairing/bonding is maintained over stm32 reboots.

I'll continue to investigate the difference between the two.

**Comment by dpgeorge** [comment]
> I can connect/disconnect/connect the phone while the stm32 is running just fine (pairing works) but after rebooting stm32 and re-connecting the pairing does not re-establish.

I'm seeing similar issues with Andriod-PYBD bonding.  Can you give a trace of the set/get secrets IRQ calls, the data (key and value) that each event is given?  (You may want to email me / obfuscate your public BD ADDR)

**Comment by andrewleech** [comment]
Yeah the secret it's trying to read upon connection is the same each time connect/disconnect/re-connect without rebooting the mpy, but then it's different after reboot mpy. 

I suspect the peer IRK isn't working / not getting re-loaded into the radio correctly. I don't think the problem is from the latest "own IRK" commit though, I was seeing the same behavior before that too. I'd say my xiaomi is likely using a static public address so resolves consistently, whereas the samsung is probably using a randomised addr so relies on the IRK. 

Tomorrow I'll trace it all through before and after the reboot to confirm the behavior and hopefully find the fix.

**Comment by dpgeorge** [comment]
> I'd say my xiaomi is likely using a static public address so resolves consistently, whereas the samsung is probably using a randomised addr so relies on the IRK.

Yes that could be it, sounds similar to what I'm seeing.

**Comment by andrewleech** [comment]
Yep just confirmed; when the phones connect to ble_bonding_peripheral.py the working Xiaomi triggers a `ble_store_read_our_sec()` with `key_sec->peer_addr.type == BLE_ADDR_PUBLIC` while the samsung has the same function with `key_sec->peer_addr.type == BLE_ADDR_RANDOM`

File: examples/bluetooth/ble_bonding_peripheral.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            return True
+        elif event == _IRQ_GET_SECRET:
+            sec_type, index, key = data
+            print("get secret:", sec_type, index, bytes(key) if key else None)
+            i = 0
+            if key is None:
+                for (t, _key), value in self._secrets.items():
+                    if t == sec_type:
+                        if i == index:
+                            return value
```
Comment (andrewleech) [inline examples/bluetooth/ble_bonding_peripheral.py:131]:
```suggestion
                        if i == index:
                            return value
                        i++                            
```

Comment (andrewleech) [reply examples/bluetooth/ble_bonding_peripheral.py:131]: @jimmo This fixes the "samsung not working over restart" issue I was having, turns out it's got nothing to do with the public/random address thing

Comment (dpgeorge) [reply examples/bluetooth/ble_bonding_peripheral.py:131]: yeah, I saw this as well

**Comment by dpgeorge** [comment]
@andrewleech does your BLE controller support LL privacy?

**Comment by andrewleech** [comment]
> @andrewleech does your BLE controller support LL privacy?
I think so, I'm using mynewt/nimble hci controller build on stm32, zephyr hci on unix. Would need to double check the build configs to be sure though.

File: examples/bluetooth/ble_bonding_peripheral.py (hunk 2 above)
Comment (dpgeorge) [reply examples/bluetooth/ble_bonding_peripheral.py:131]: this is now fixed in #6662

**Comment by dpgeorge** [comment]
> Write multi-test.

Two basic multitests for pairing and bonding have been added in #6662. 

> PYBD doesn't work with RPA mode.

The BT controller on PYBD doesn't support LL privacy in the controller so RPA would need to be added on the host side.

**Comment by dpgeorge** [comment]
I've tested this and am happy with how it works.

Merged in f2a9a0ac413cc337ccd78bbb26cd67cee7b1e210 through d79b9c6c7cb42c69c15d1b7869a04e889d812b82

Outstanding items can be addressed in follow-up work.

Thank you @jimmo and @andrewleech !

**Pull request closed without merging**